### ISSUE TITLE
handle failures for corner cases if recurse=True in dump

### ISSUE THREAD

**Issue by mmckerns** [uqfoundation/dill#105]
If certain corner cases can be handled, then `recurse=True` could be the default setting for `dump`, which would be more robust.

Failure with `recurse=True` noted in some tests: (`test_classdef` and others).

**Comment by mmckerns** [comment]
due to merge of edited code from #86

**Comment by rgbkrk** [comment]
Nice, thanks.

**Comment by mmckerns** [comment]
Test failures with `recurse=True` (by modifying `settings['recurse']` inside of test file).
```
dude@hilbert>$ python test_classdef.py 
Traceback (most recent call last):
  File "test_classdef.py", line 79, in <module>
    assert _obj.ok()
  File "test_classdef.py", line 17, in ok
    return True
NameError: global name 'True' is not defined
```
```
dude@hilbert>$ python test_mixins.py 
Traceback (most recent call last):
  File "test_mixins.py", line 75, in <module>
    assert _d(1,2,3) == -2*fx
  File "test_mixins.py", line 26, in func
    x = f(*args, **kwds)
  File "test_mixins.py", line 37, in double_add
    return sum(args)
NameError: global name 'sum' is not defined
```
```
dude@hilbert>$ python test_module.py 
Traceback (most recent call last):
  File "test_module.py", line 30, in <module>
    assert module.double_add(1, 2, 3) == 2 * module.fx
  File "/Users/mmckerns/dev/svn/pathos/dill/tests/test_mixins.py", line 26, in func
    x = f(*args, **kwds)
  File "/Users/mmckerns/dev/svn/pathos/dill/tests/test_mixins.py", line 37, in double_add
    return sum(args)
NameError: global name 'sum' is not defined
```
```
dude@hilbert>$ python test_nested.py 
Traceback (most recent call last):
  File "test_nested.py", line 104, in <module>
    assert error is None
AssertionError
```
```
dude@hilbert>$ python test_moduledict.py
Traceback (most recent call last):
  File "test_moduledict.py", line 20, in <module>
    assert dill.pickles(f2)
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.3.dev0-py2.7.egg/dill/dill.py", line 1094, in pickles
    pik = copy(obj, **kwds)
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.3.dev0-py2.7.egg/dill/dill.py", line 159, in copy
    return loads(dumps(obj, *args, **kwds))
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.3.dev0-py2.7.egg/dill/dill.py", line 192, in dumps
    dump(obj, file, protocol, byref, fmode, recurse)#, strictio)
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.3.dev0-py2.7.egg/dill/dill.py", line 186, in dump
    pik.dump(obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 224, in dump
    self.save(obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.3.dev0-py2.7.egg/dill/dill.py", line 693, in save_function
    globs = globalvars(obj, recurse=True)
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.3.dev0-py2.7.egg/dill/detect.py", line 190, in globalvars
    func.update(globalvars(globs.get(key), recurse=True).keys())
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.3.dev0-py2.7.egg/dill/detect.py", line 190, in globalvars
    func.update(globalvars(globs.get(key), recurse=True).keys())
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.3.dev0-py2.7.egg/dill/detect.py", line 190, in globalvars
    func.update(globalvars(globs.get(key), recurse=True).keys())
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.3.dev0-py2.7.egg/dill/detect.py", line 190, in globalvars
    func.update(globalvars(globs.get(key), recurse=True).keys())
  …
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.3.dev0-py2.7.egg/dill/detect.py", line 190, in globalvars
    func.update(globalvars(globs.get(key), recurse=True).keys())
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.3.dev0-py2.7.egg/dill/detect.py", line 190, in globalvars
    func.update(globalvars(globs.get(key), recurse=True).keys())
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.3.dev0-py2.7.egg/dill/detect.py", line 187, in globalvars
    func = nestedglobals(getattr(func, func_code)) #, depth=None)
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.3.dev0-py2.7.egg/dill/detect.py", line 162, in nestedglobals
    names |= nestedglobals(co, recurse=True)
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.3.dev0-py2.7.egg/dill/detect.py", line 154, in nestedglobals
    with capture('stdout') as out:
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.3.dev0-py2.7.egg/dill/temp.py", line 40, in capture
    setattr(sys, stream, StringIO())
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/StringIO.py", line 56, in __init__
    if not isinstance(buf, basestring):
RuntimeError: maximum recursion depth exceeded while calling a Python object
```

**Comment by mmckerns** [comment]
The issue with `test_classdef` is that `True` is not found by `globalvars`.  This can be fixed by adding:
 ```
globs = {} #FIXME: vars(getmodule(object)) # get dict of __builtins__
```
at https://github.com/uqfoundation/dill/blob/master/dill/detect.py#L184.

However, the above affects the imports returned from `dill.source.likely_import` (and friends) -- there appears to be some mixing of `__builtin__` and `__main__`.
 
Also were the above changed,`globalvars` still fails to find globals referred to inside of classes.

**Comment by mmckerns** [comment]
Note that `multiprocess` and `pathos` perform much more robustly on Windows when `dill.settings['recurse'] = True`.

See: https://github.com/uqfoundation/multiprocess/issues/6

**Comment by mmckerns** [comment]
Note discussion here: #115.

**Comment by mmckerns** [comment]
4a66a75f7bacf00872fed1596566b8c6b37315a5 fixes all except `test_moduledict.py` recursion issue

**Comment by mmckerns** [comment]
@mrocklin: only one corner case left to address, then will most likely use `recurse` as the default.

**Comment by mrocklin** [comment]
I'm very glad to hear it.

**Comment by mmckerns** [comment]
I'll also be cutting a release immediately afterward.

**Comment by mmckerns** [comment]
final corner case of recursive referencing handled (somewhat hackishly) in 758808d9a41c95999bc898aeff9b44c742e0da43.

**Comment by mmckerns** [comment]
Reopening… other corner cases discovered for different versions of python.

`python2.7`: all tests pass

`python2.5` and `python3.2` and `python3.3`: 
```
$ python test_classdef.py 
Traceback (most recent call last):
  File "test_classdef.py", line 116, in <module>
    assert dill.pickles(a1)
AssertionError
```

`python3.1` and `python3.2` and `python3.3`: 
```
$ python test_mixins.py 
Traceback (most recent call last):
  File "test_mixins.py", line 74, in <module>
    _d = dill.copy(double_add)
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 159, in copy
    return loads(dumps(obj, *args, **kwds))
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 192, in dumps
    dump(obj, file, protocol, byref, fmode, recurse)#, strictio)
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 186, in dump
    pik.dump(obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 237, in dump
    self.save(obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 708, in save_function
    obj.__dict__), obj=obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 414, in save_reduce
    save(args)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 543, in save_tuple
    save(element)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 543, in save_tuple
    save(element)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 917, in save_cell
    pickler.save_reduce(_create_cell, (obj.cell_contents,), obj=obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 414, in save_reduce
    save(args)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 529, in save_tuple
    save(element)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 708, in save_function
    obj.__dict__), obj=obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 414, in save_reduce
    save(args)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 543, in save_tuple
    save(element)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 745, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 623, in save_dict
    self._batch_setitems(obj.items())
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 661, in _batch_setitems
    save(v)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 872, in save_builtin_method
    pickler.save_reduce(_get_attr, (obj.__self__, obj.__name__), obj=obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 418, in save_reduce
    self.memoize(obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 257, in memoize
    assert id(obj) not in self.memo
AssertionError
```
```
$ python test_module.py 
Traceback (most recent call last):
  File "test_module.py", line 20, in <module>
    pik_mod = dill.dumps(module)
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 192, in dumps
    dump(obj, file, protocol, byref, fmode, recurse)#, strictio)
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 186, in dump
    pik.dump(obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 237, in dump
    self.save(obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 1046, in save_module
    state=_main_dict)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 432, in save_reduce
    save(state)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 745, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 623, in save_dict
    self._batch_setitems(obj.items())
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 656, in _batch_setitems
    save(v)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 708, in save_function
    obj.__dict__), obj=obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 414, in save_reduce
    save(args)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 543, in save_tuple
    save(element)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 543, in save_tuple
    save(element)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 917, in save_cell
    pickler.save_reduce(_create_cell, (obj.cell_contents,), obj=obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 414, in save_reduce
    save(args)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 529, in save_tuple
    save(element)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 708, in save_function
    obj.__dict__), obj=obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 414, in save_reduce
    save(args)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 543, in save_tuple
    save(element)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 745, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 623, in save_dict
    self._batch_setitems(obj.items())
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 661, in _batch_setitems
    save(v)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 299, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 872, in save_builtin_method
    pickler.save_reduce(_get_attr, (obj.__self__, obj.__name__), obj=obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 418, in save_reduce
    self.memoize(obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 257, in memoize
    assert id(obj) not in self.memo
AssertionError
```

`python3.4`:
```
$ python test_classdef.py 
(<class 'AttributeError'>, AttributeError("'module' object has no attribute 'globals'",), <traceback object at 0x10fce11c8>)
<function _get_attr at 0x10efa86a8> (<module '__main__' from 'test_classdef.py'>, 'globals')
Traceback (most recent call last):
  File "test_classdef.py", line 116, in <module>
    assert dill.pickles(a1)
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 1106, in pickles
    pik = copy(obj, **kwds)
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 159, in copy
    return loads(dumps(obj, *args, **kwds))
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 209, in loads
    return load(file)
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 199, in load
    obj = pik.load()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/pickle.py", line 1038, in load
    dispatch[key[0]](self)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/pickle.py", line 1384, in load_reduce
    value = func(*args)
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 658, in _get_attr
    return getattr(self, name)
AttributeError: 'module' object has no attribute 'globals'
```
```
$ python test_mixins.py 
(<class 'AttributeError'>, AttributeError("'module' object has no attribute '__import__'",), <traceback object at 0x1033d0b88>)
<function _get_attr at 0x1026c46a8> (<module '__main__' from 'test_mixins.py'>, '__import__')
Traceback (most recent call last):
  File "test_mixins.py", line 74, in <module>
    _d = dill.copy(double_add)
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 159, in copy
    return loads(dumps(obj, *args, **kwds))
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 209, in loads
    return load(file)
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 199, in load
    obj = pik.load()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/pickle.py", line 1038, in load
    dispatch[key[0]](self)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/pickle.py", line 1384, in load_reduce
    value = func(*args)
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 658, in _get_attr
    return getattr(self, name)
AttributeError: 'module' object has no attribute '__import__'
```
```
$ python test_module.py 
(<class 'AttributeError'>, AttributeError("'module' object has no attribute 'max'",), <traceback object at 0x103422608>)
<function _get_attr at 0x1027166a8> (<module 'test_mixins' from '/Users/mmckerns/dev/svn/pathos/dill/tests/test_mixins.py'>, 'max')
Traceback (most recent call last):
  File "test_module.py", line 28, in <module>
    module = dill.loads(pik_mod)
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 209, in loads
    return load(file)
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 199, in load
    obj = pik.load()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/pickle.py", line 1038, in load
    dispatch[key[0]](self)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/pickle.py", line 1384, in load_reduce
    value = func(*args)
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 658, in _get_attr
    return getattr(self, name)
AttributeError: 'module' object has no attribute 'max'
```
```

```

**Comment by mmckerns** [comment]
Also, for `python3.4` and similarly for other `3.x`:
```
$ python test_objects.py 
(<class 'AttributeError'>, AttributeError("'module' object has no attribute 'iter'",), <traceback object at 0x10570d8c8>)
<function _get_attr at 0x1049d66a8> (<module 'weakref' from '/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/weakref.py'>, 'iter')
fails: IzipType <class 'zip'>
(<class 'AttributeError'>, AttributeError("'module' object has no attribute 'iter'",), <traceback object at 0x10570d348>)
<function _get_attr at 0x1049d66a8> (<module 'weakref' from '/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/weakref.py'>, 'iter')
fails: ChainType <class 'itertools.chain'>
fails: BZ2FileType <class 'bz2.BZ2File'>
fails: BZ2CompressorType <class '_bz2.BZ2Compressor'>
fails: BZ2DecompressorType <class '_bz2.BZ2Decompressor'>
fails: SocketType <class 'socket.socket'>
fails: SocketPairType <class 'socket.socket'>
(<class 'AttributeError'>, AttributeError("'module' object has no attribute 'max'",), <traceback object at 0x105701108>)
<function _get_attr at 0x1049d66a8> (<module 'weakref' from '/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/weakref.py'>, 'max')
fails: GeneratorContextManagerType <class 'contextlib._GeneratorContextManager'>
fails: FileType <class '_io.FileIO'>
(<class 'AttributeError'>, AttributeError("'module' object has no attribute 'iter'",), <traceback object at 0x105716b88>)
<function _get_attr at 0x1049d66a8> (<module '_weakrefset' from '/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/_weakrefset.py'>, 'iter')
fails: ListIteratorType <class 'list_iterator'>
(<class 'AttributeError'>, AttributeError("'module' object has no attribute 'iter'",), <traceback object at 0x105716b88>)
<function _get_attr at 0x1049d66a8> (<module '_weakrefset' from '/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/_weakrefset.py'>, 'iter')
fails: TupleIteratorType <class 'tuple_iterator'>
(<class 'AttributeError'>, AttributeError("'module' object has no attribute 'iter'",), <traceback object at 0x105716948>)
<function _get_attr at 0x1049d66a8> (<module '_weakrefset' from '/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/_weakrefset.py'>, 'iter')
fails: XRangeIteratorType <class 'range_iterator'>
(<class 'AttributeError'>, AttributeError("'module' object has no attribute 'iter'",), <traceback object at 0x1057072c8>)
<function _get_attr at 0x1049d66a8> (<module '_weakrefset' from '/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/_weakrefset.py'>, 'iter')
fails: CycleType <class 'itertools.cycle'>
fails: GzipFileType <class 'gzip.GzipFile'>
(<class 'AttributeError'>, AttributeError("'module' object has no attribute 'iter'",), <traceback object at 0x10618dec8>)
<function _get_attr at 0x1049d66a8> (<module '_weakrefset' from '/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/_weakrefset.py'>, 'iter')
fails: CompressType <class 'itertools.compress'>
(<class 'AttributeError'>, AttributeError("'module' object has no attribute 'getattr'",), <traceback object at 0x10618ddc8>)
<function _get_attr at 0x1049d66a8> (<module '_weakrefset' from '/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/_weakrefset.py'>, 'getattr')
fails: MethodWrapperType <class 'method-wrapper'>
```

**Comment by mmckerns** [comment]
`test_objects.py` fixed by #118.

**Comment by mmckerns** [comment]
three failing tests for `python3.x` also fixed by #118.

**Comment by mmckerns** [comment]
all remaining tests fixed by #118.

**Comment by mmckerns** [comment]
after #118, looking at remaining errors:

`python2.5` to `python2.7`:
```
$ python test_classdef.py 
Traceback (most recent call last):
  File "test_classdef.py", line 79, in <module>
    assert _obj.ok()
  File "test_classdef.py", line 17, in ok
    return True
NameError: global name 'True' is not defined
```
`python2.5` to `python3.4`
```
$ python test_mixins.py 
Traceback (most recent call last):
  File "test_mixins.py", line 75, in <module>
    assert _d(1,2,3) == -2*fx
  File "test_mixins.py", line 26, in func
    x = f(*args, **kwds)
  File "test_mixins.py", line 37, in double_add
    return sum(args)
NameError: global name 'sum' is not defined
```
```
$ python test_module.py 
Traceback (most recent call last):
  File "test_module.py", line 30, in <module>
    assert module.double_add(1, 2, 3) == 2 * module.fx
  File "/Users/mmckerns/dev/svn/pathos/dill/tests/test_mixins.py", line 26, in func
    x = f(*args, **kwds)
  File "/Users/mmckerns/dev/svn/pathos/dill/tests/test_mixins.py", line 37, in double_add
    return sum(args)
NameError: global name 'sum' is not defined
```
`python2.5` to `python3.4` (similar errors)
```
$ python test_moduledict.py 
Traceback (most recent call last):
  File "test_moduledict.py", line 20, in <module>
    assert dill.pickles(f2)
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.4.dev0-py2.7.egg/dill/dill.py", line 1107, in pickles
    pik = copy(obj, **kwds)
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.4.dev0-py2.7.egg/dill/dill.py", line 159, in copy
    return loads(dumps(obj, *args, **kwds))
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.4.dev0-py2.7.egg/dill/dill.py", line 192, in dumps
    dump(obj, file, protocol, byref, fmode, recurse)#, strictio)
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.4.dev0-py2.7.egg/dill/dill.py", line 186, in dump
    pik.dump(obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 224, in dump
    self.save(obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.4.dev0-py2.7.egg/dill/dill.py", line 714, in save_function
    obj.__dict__), obj=obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 401, in save_reduce
    save(args)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 562, in save_tuple
    save(element)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.4.dev0-py2.7.egg/dill/dill.py", line 746, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 649, in save_dict
    self._batch_setitems(obj.iteritems())
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 686, in _batch_setitems
    save(v)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.4.dev0-py2.7.egg/dill/dill.py", line 714, in save_function
    obj.__dict__), obj=obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 401, in save_reduce
    save(args)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 562, in save_tuple
    save(element)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.4.dev0-py2.7.egg/dill/dill.py", line 746, in save_module_dict
    StockPickler.save_dict(pickler, obj)
...
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 649, in save_dict
    self._batch_setitems(obj.iteritems())
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 686, in _batch_setitems
    save(v)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.4.dev0-py2.7.egg/dill/dill.py", line 702, in save_function
    globs = globalvars(obj, recurse=True)
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.4.dev0-py2.7.egg/dill/detect.py", line 200, in globalvars
    func = set(nestedglobals(getattr(func, func_code)))
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.4.dev0-py2.7.egg/dill/detect.py", line 165, in nestedglobals
    with capture('stdout') as out:
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.4.dev0-py2.7.egg/dill/temp.py", line 40, in capture
    setattr(sys, stream, StringIO())
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/StringIO.py", line 56, in __init__
    if not isinstance(buf, basestring):
RuntimeError: maximum recursion depth exceeded in __instancecheck__ 
```

**Comment by mmckerns** [comment]
after reapplying a few lost lines of code… the error state is returned to -15 hours ago (just before #118 was applied)

**Comment by matsjoyce** [comment]
Sorry! :tired_face:

**Comment by mmckerns** [comment]
@matsjoyce: no worries. I missed it as well.  It's reverted and merged.

**Comment by mmckerns** [comment]
All errors (not the serialization failures) in `test_objects.py` stem from serialization of certain`weakref` objects -- added as it's own ticket (see #119).

**Comment by mmckerns** [comment]
The `test_module` failure seems to be purely related to `test_mixins` behavior, so it should be solved by solving `test_mixins`… opened a new ticket for this in #120

**Comment by mmckerns** [comment]
Current test failures:

`python2.5`, `python3.2`, `python3.3`:
```
$ python test_classdef.py 
Traceback (most recent call last):
  File "test_classdef.py", line 116, in <module>
    assert dill.pickles(a1)
AssertionError
```

`python3.1`, `python3.2`, `python3.3`:   (as before)
```
$ python test_mixins.py 
Traceback (most recent call last):
  File "test_mixins.py", line 74, in <module>
    _d = dill.copy(double_add)
  File "/Users/mmckerns/lib/python3.1/site-packages/dill-0.2.4.dev0-py3.1.egg/dill/dill.py", line 159, in copy
...
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.1/lib/python3.1/pickle.py", line 257, in memoize
    assert id(obj) not in self.memo
AssertionError
```
similarly for `$ python test_module.py`

`python3.4`:
```
$ python test_classdef.py 
Traceback (most recent call last):
  File "test_classdef.py", line 117, in <module>
    assert a1.__dict__ == dill.copy(a1).__dict__
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 159, in copy
    return loads(dumps(obj, *args, **kwds))
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 209, in loads
    return load(file)
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 199, in load
    obj = pik.load()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/pickle.py", line 1038, in load
    dispatch[key[0]](self)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/pickle.py", line 1480, in load_setitems
    dict[stack[i]] = stack[i + 1]
TypeError: unhashable type: 'list'
```
```
$ python test_mixins.py 
Traceback (most recent call last):
  File "test_mixins.py", line 86, in <module>
    assert ds.importable(quadish, source=False) == 'from %s import quadish\n' % __name__
AssertionError
```

**Comment by mmckerns** [comment]
Interesting… the `python3.4` failure for `test_classdef` only occurs when the `TestArray` instance is pickled **more than once** *and*  `test_classdef` is run as `__main__` (i.e. `TestArray` is not in the interpreter or imported).

**Comment by mmckerns** [comment]
`python3.4` issue for `test_classdef.py` solved by 81cd9f838318516269abcc74fc27381fc36c4d7b.
Errors for `python2.5`, and `python3.1`-`python3.3` remain the same.  Error for `test_mixins.py` for `python3.4` is now:
```
$ python test_mixins.py 
Traceback (most recent call last):
  File "test_mixins.py", line 102, in <module>
    assert result == 'def quad(a=1, b=1, c=0):\n  inverted = [False]\n  def invert():\n    inverted[0] = not inverted[0]\n  def dec(f):\n    def func(*args, **kwds):\n      x = f(*args, **kwds)\n      if inverted[0]: x = -x\n      return a*x**2 + b*x + c\n    func.__wrapped__ = f\n    func.invert = invert\n    return func\n  return dec\n\n@quad(a=0,b=2)\ndef double_add(*args):\n  return sum(args)\n'
AssertionError
```

**Comment by mmckerns** [comment]
fix for `test_mixins.py` for `python3.4` in 5fe8c677a2c905fbb8aa660c1d16f6162fd826fa 

The remaining two Errors are not affected.

**Comment by mmckerns** [comment]
Looks like the issue with `test_mixins.py` is that `recurse=True` takes a module as a dependency, and then all the dependencies in the module (i.e. all `vars`) -- eventually hitting something unpicklable.  The solution might be to simply always disable `recurse` for certain objects.

**Comment by mmckerns** [comment]
Minimal example (`python3.3`):
```
# test.py
import dill
dill.detect.trace(True)
dill.settings['recurse'] = True

from math import sin

def foo(x):
  return sin(x)

dill.copy(foo)
```
```
$ python test.py 
F1: <function foo at 0x10dfdbc20>
F2: <function _create_function at 0x10ef217a0>
Co: <code object foo at 0x10df711e0, file "test.py", line 7>
F2: <function _unmarshal at 0x10ef215f0>
D2: <dict object at 0x10e2e58c0>
B1: <built-in function sin>
F2: <function _get_attr at 0x10eea50e0>
M2: <module 'math' from '/opt/local/Library/Frameworks/Python.framework/Versions/3.3/lib/python3.3/lib-dynload/math.so'>
F2: <function _import_module at 0x10eea5200>
D2: <dict object at 0x10e258050>
```

Now, edited to use `sum` instead of `sin`:
```
$ python test.py 
F1: <function foo at 0x100573c20>
F2: <function _create_function at 0x1014b27a0>
Co: <code object foo at 0x1005091e0, file "test.py", line 5>
F2: <function _unmarshal at 0x1014b25f0>
D2: <dict object at 0x1008778c0>
B1: <built-in function sum>
F2: <function _get_attr at 0x1014360e0>
M1: <module '__main__' from 'test.py'>
F2: <function _import_module at 0x101436200>
D2: <dict object at 0x1005cf248>
B1: <built-in function __build_class__>
T1: <class 'ConnectionResetError'>
F2: <function _load_type at 0x1014b2680>
T1: <class 'complex'>
B1: <built-in function issubclass>
T1: <class 'BlockingIOError'>
B1: <built-in function compile>
B1: <built-in function sum>
T1: <class 'zip'>
T1: <class 'OSError'>
T1: <class 'site.setquit.<locals>.Quitter'>
D2: <dict object at 0x1006abcf8>
T1: <class 'PendingDeprecationWarning'>
T1: <class 'UserWarning'>
T1: <class 'NotADirectoryError'>
T1: <class 'ImportError'>
B1: <built-in function delattr>
B1: <built-in function hasattr>
T1: <class 'ConnectionRefusedError'>
B1: <built-in function isinstance>
B1: <built-in function sorted>
T1: <class 'str'>
T1: <class 'staticmethod'>
B1: <built-in function abs>
T1: <class 'range'>
T1: <class 'UnicodeTranslateError'>
T1: <class 'UnicodeWarning'>
B1: <built-in function min>
T1: <class 'NameError'>
T1: <class 'map'>
T1: <class 'AssertionError'>
Si: NotImplemented
F2: <function _eval_repr at 0x1014b2ef0>
T1: <class 'OverflowError'>
T1: <class 'type'>
T1: <class 'BytesWarning'>
T1: <class 'PermissionError'>
B1: <built-in function repr>
B1: <built-in function exec>
T4: <class 'site._Printer'>
D2: <dict object at 0x100785a70>
B1: <built-in function dir>
T1: <class 'ReferenceError'>
B1: <built-in function open>
M2: <module 'io' (built-in)>
Si: Ellipsis
B1: <built-in function eval>
T1: <class 'filter'>
D2: <dict object at 0x1007a0d40>
B1: <built-in function format>
T1: <class 'TimeoutError'>
T1: <class 'bytearray'>
T1: <class 'ValueError'>
T1: <class 'ProcessLookupError'>
B1: <built-in function next>
T1: <class 'list'>
T1: <class 'object'>
T1: <class 'KeyboardInterrupt'>
T1: <class 'classmethod'>
T4: <class 'site._Helper'>
T1: <class 'EOFError'>
T1: <class 'slice'>
B1: <built-in function bin>
T1: <class 'BufferError'>
B1: <built-in function oct>
T1: <class 'property'>
B1: <built-in function print>
T1: <class 'frozenset'>
B1: <built-in function divmod>
T1: <class 'bool'>
B1: <built-in function chr>
T1: <class 'int'>
T1: <class 'StopIteration'>
T1: <class 'Exception'>
T1: <class 'ImportWarning'>
T1: <class 'FutureWarning'>
B1: <built-in function id>
T1: <class 'IndexError'>
B1: <built-in function ord>
T1: <class 'memoryview'>
B1: <built-in function __import__>
T1: <class 'super'>
T1: <class 'NotImplementedError'>
T1: <class 'UnicodeEncodeError'>
T1: <class 'BrokenPipeError'>
T1: <class 'AttributeError'>
T1: <class 'FileExistsError'>
T1: <class 'IsADirectoryError'>
T1: <class 'TypeError'>
T1: <class 'SystemExit'>
T1: <class 'BaseException'>
T1: <class 'LookupError'>
B1: <built-in function input>
T1: <class 'ConnectionAbortedError'>
T1: <class 'Warning'>
T1: <class 'ArithmeticError'>
B1: <built-in function callable>
B1: <built-in function len>
B1: <built-in function hex>
B1: <built-in function hash>
T1: <class 'FileNotFoundError'>
T1: <class 'tuple'>
T1: <class 'RuntimeWarning'>
B1: <built-in function any>
B1: <built-in function ascii>
D2: <dict object at 0x1007a0758>
T1: <class 'enumerate'>
T1: <class 'bytes'>
B1: <built-in function pow>
T1: <class 'ChildProcessError'>
B1: <built-in function vars>
T1: <class 'UnicodeError'>
T1: <class 'DeprecationWarning'>
T1: <class 'reversed'>
B1: <built-in function locals>
T1: <class 'SystemError'>
T1: <class 'UnicodeDecodeError'>
D2: <dict object at 0x1007855f0>
B1: <built-in function round>
T1: <class 'ZeroDivisionError'>
T1: <class 'UnboundLocalError'>
T1: <class 'TabError'>
T1: <class 'float'>
T1: <class 'RuntimeError'>
T1: <class 'MemoryError'>
M2: <module 'dill' from '/Users/mmckerns/lib/python3.3/site-packages/dill-0.2.4.dev0-py3.3.egg/dill/__init__.py'>
B1: <built-in function globals>
B1: <built-in function max>
T1: <class 'SyntaxWarning'>
T1: <class 'KeyError'>
T1: <class 'SyntaxError'>
T1: <class 'FloatingPointError'>
F1: <function foo at 0x100573c20>
D1: <dict object at 0x1005376c8>
D2: <dict object at 0x1007eb050>
T1: <class 'ConnectionError'>
T1: <class 'GeneratorExit'>
T1: <class 'IndentationError'>
T1: <class 'ResourceWarning'>
T1: <class 'dict'>
B1: <built-in function getattr>
B1: <built-in function setattr>
B1: <built-in function iter>
T1: <class 'InterruptedError'>
T1: <class 'set'>
B1: <built-in function all>
Traceback (most recent call last):
  File "test.py", line 8, in <module>
    dill.copy(foo)
…
  File "/Users/mmckerns/lib/python3.3/site-packages/dill-0.2.4.dev0-py3.3.egg/dill/dill.py", line 877, in save_builtin_method
    pickler.save_reduce(_get_attr, (obj.__self__, obj.__name__), obj=obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.3/lib/python3.3/pickle.py", line 416, in save_reduce
    self.memoize(obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.3/lib/python3.3/pickle.py", line 255, in memoize
    assert id(obj) not in self.memo
AssertionError
```

So the issue appears to be when a function from `builtins` is referred to as a `global`, then `recurse=True` tries to pickle every object in `builtins` individually… until it hits an unpicklable object.  In `python2.7` and `python3.4`, it succeeds only because `dill` pickles all the objects in `builtins`.

`python3.4`:
```
$ python test.py 
F1: <function foo at 0x105221d08>
F2: <function _create_function at 0x105597d90>
Co: <code object foo at 0x105251e40, file "test.py", line 5>
F2: <function _unmarshal at 0x105597bf8>
D2: <dict object at 0x10626e648>
B1: <built-in function sum>
F2: <function _get_attr at 0x10559d6a8>
M1: <module '__main__' from 'test.py'>
F2: <function _import_module at 0x10559d7b8>
D2: <dict object at 0x10529f148>
T1: <class 'Exception'>
F2: <function _load_type at 0x105597c80>
B1: <built-in function min>
T1: <class 'ConnectionRefusedError'>
T1: <class 'EOFError'>
T1: <class 'zip'>
F1: <function foo at 0x105221d08>
D1: <dict object at 0x105232788>
D2: <dict object at 0x1055a7208>
T1: <class 'TypeError'>
T1: <class 'type'>
T1: <class 'staticmethod'>
B1: <built-in function hex>
B1: <built-in function id>
T1: <class 'OverflowError'>
T1: <class 'AssertionError'>
B1: <built-in function callable>
T1: <class 'DeprecationWarning'>
B1: <built-in function len>
B1: <built-in function getattr>
T4: <class '_sitebuiltins._Printer'>
D2: <dict object at 0x1054762c8>
B1: <built-in function vars>
B1: <built-in function iter>
B1: <built-in function isinstance>
B1: <built-in function __import__>
T1: <class 'FutureWarning'>
T1: <class 'str'>
T1: <class 'InterruptedError'>
T1: <class 'FileNotFoundError'>
T1: <class 'property'>
T1: <class 'KeyboardInterrupt'>
T1: <class 'object'>
B1: <built-in function round>
T1: <class 'ReferenceError'>
T1: <class 'ConnectionAbortedError'>
D2: <dict object at 0x10545a048>
B1: <built-in function chr>
T1: <class 'StopIteration'>
T1: <class 'memoryview'>
T1: <class 'NotADirectoryError'>
B1: <built-in function input>
T1: <class 'bool'>
T1: <class 'OSError'>
T1: <class 'ImportWarning'>
T1: <class 'UnicodeTranslateError'>
T1: <class 'ImportError'>
B1: <built-in function max>
T1: <class '_sitebuiltins.Quitter'>
D2: <dict object at 0x105476ec8>
T1: <class 'LookupError'>
T1: <class 'UnicodeEncodeError'>
T1: <class 'UnicodeError'>
T1: <class 'list'>
T1: <class 'bytearray'>
B1: <built-in function compile>
T1: <class 'UnicodeWarning'>
T1: <class 'ConnectionResetError'>
T1: <class 'map'>
B1: <built-in function pow>
B1: <built-in function exec>
T1: <class 'BlockingIOError'>
T1: <class 'dict'>
B1: <built-in function locals>
T1: <class 'BaseException'>
T1: <class 'SyntaxError'>
T1: <class 'UnboundLocalError'>
B1: <built-in function print>
Si: Ellipsis
F2: <function _eval_repr at 0x10559d510>
M2: <module 'dill' from '/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/__init__.py'>
T1: <class 'classmethod'>
T1: <class 'IsADirectoryError'>
B1: <built-in function issubclass>
T1: <class 'ArithmeticError'>
T1: <class 'BytesWarning'>
T1: <class 'GeneratorExit'>
B1: <built-in function oct>
B1: <built-in function eval>
T1: <class 'RuntimeError'>
B1: <built-in function globals>
B1: <built-in function next>
T1: <class 'FileExistsError'>
T1: <class 'BufferError'>
T1: <class 'NotImplementedError'>
T1: <class 'int'>
T1: <class 'set'>
B1: <built-in function delattr>
T1: <class 'ResourceWarning'>
T1: <class 'SystemError'>
T1: <class 'UnicodeDecodeError'>
T1: <class 'bytes'>
B1: <built-in function ascii>
T4: <class '_sitebuiltins._Helper'>
T1: <class 'BrokenPipeError'>
B1: <built-in function repr>
B1: <built-in function __build_class__>
T1: <class 'slice'>
D2: <dict object at 0x105476e88>
B1: <built-in function setattr>
T1: <class 'KeyError'>
T1: <class 'PermissionError'>
T1: <class 'TabError'>
T1: <class 'filter'>
B1: <built-in function sorted>
T1: <class 'FloatingPointError'>
B1: <built-in function dir>
T1: <class 'super'>
T1: <class 'IndexError'>
T1: <class 'RuntimeWarning'>
B1: <built-in function abs>
T1: <class 'MemoryError'>
T1: <class 'tuple'>
T1: <class 'NameError'>
T1: <class 'float'>
B1: <built-in function ord>
Si: NotImplemented
B1: <built-in function sum>
T1: <class 'reversed'>
T1: <class 'SyntaxWarning'>
B1: <built-in function format>
T1: <class 'SystemExit'>
T1: <class 'Warning'>
B1: <built-in function all>
B1: <built-in function hash>
T1: <class 'AttributeError'>
T1: <class 'enumerate'>
T1: <class 'ProcessLookupError'>
T1: <class 'ConnectionError'>
T1: <class 'range'>
T1: <class 'ValueError'>
T1: <class 'ZeroDivisionError'>
T1: <class 'UserWarning'>
B1: <built-in function open>
M2: <module 'io' (built-in)>
B1: <built-in function divmod>
T1: <class 'frozenset'>
B1: <built-in function any>
B1: <built-in function bin>
T1: <class 'complex'>
T1: <class 'PendingDeprecationWarning'>
B1: <built-in function hasattr>
T1: <class 'TimeoutError'>
D2: <dict object at 0x1054805c8>
T1: <class 'IndentationError'>
T1: <class 'ChildProcessError'>
```

**Comment by mmckerns** [comment]
Note also `builtins` is misidentified as `__main__`.   This may be the key issue.  This behavior was (I believe) observed in generating `likely_import` for `sum` (although the details are lost…), and in part was the reason for adding the lookup in `builtins` in `dill.dill._get_attr`.

**Comment by mmckerns** [comment]
all above errors (except `test_classdef.py` in `python2.5`) fixed by 1938291419dc48c3b0930774b2c18272ccc3b1a4.  Will open new issue for this last case.
